phoenix-core/src/main/java/org/apache/phoenix/index/GlobalIndexChecker.java

```diff
@@ -0,0 +1,384 @@
+                do {
+                    hasMore = scanner.nextRaw(result);
+                    if (result.isEmpty()) {
+                        break;
+                    }
+                    if (verifyRowAndRepairIfNecessary(result)) {
+                        break;
+                    }
+                    // skip this row as it is invalid
+                    // if there is no more row, then result will be an empty list
+                } while (hasMore);
+                return hasMore;
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(region.getRegionInfo().getRegionNameAsString(), t);
+                return false; // impossible
+            }
+        }
+
+        private void deleteRowIfAgedEnough(byte[] indexRowKey, long ts) throws IOException {
+            if ((EnvironmentEdgeManager.currentTimeMillis() - ts) > ageThreshold) {
+                Delete del = new Delete(indexRowKey, ts);
+                Mutation[] mutations = new Mutation[]{del};
+                region.batchMutate(mutations);
+            }
+        }
+
+        private void repairIndexRows(byte[] indexRowKey, long ts, List<Cell> row) throws IOException {
+            // Build the data table row key from the index table row key
+            if (buildIndexScan == null) {
+                buildIndexScan = new Scan();
+                indexScan = new Scan(scan);
+                byte[] dataTableName = scan.getAttribute(PHYSICAL_DATA_TABLE_NAME);
+                byte[] indexTableName = region.getRegionInfo().getTable().getName();
+                dataHTable = ServerUtil.getHTableForCoprocessorScan(env,
+                        SchemaUtil.getPhysicalTableName(dataTableName, env.getConfiguration()));
+                if (indexMaintainer == null) {
+                    byte[] md = scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD);
+                    List<IndexMaintainer> maintainers = IndexMaintainer.deserialize(md, true);
+                    indexMaintainer = getIndexMaintainer(maintainers, indexTableName);
+                }
+                if (indexMaintainer == null) {
+                    throw new DoNotRetryIOException(
+                            "repairIndexRows: IndexMaintainer is not included in scan attributes for " +
+                                    region.getRegionInfo().getTable().getNameAsString());
+                }
+                if (viewConstants == null) {
+                    viewConstants = IndexUtil.deserializeViewConstantsFromScan(scan);
+                }
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.UNGROUPED_AGG, TRUE_BYTES);
+                buildIndexScan.setAttribute(PhoenixIndexCodec.INDEX_PROTO_MD, scan.getAttribute(PhoenixIndexCodec.INDEX_PROTO_MD));
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.REBUILD_INDEXES, TRUE_BYTES);
+                buildIndexScan.setAttribute(BaseScannerRegionObserver.SCAN_LIMIT, Bytes.toBytes(repairCount));
```

Comment:
why do we rebuild more than one row here?  My understanding is at this point, we've found one unverified row so far as we enter this method?

Reply:
Building one row or a batch if 100 rows has the same cost in practice. Since Phoenix does table updates in batches, I expect that will have batches of unverified rows instead of just one. Here we are doing "repair ahead" for this reason.